### DISCUSSION TITLE
Possible Workaround to Obtain Most-up-to-date Residuals of StressDivergenceTensor Kernels

### DISCUSSION THREAD

**Discussion by chunhuizhao478** [idaholab/moose#22841] [Q&A Modules: Solid mechanics] [answered]
Hi all,

I have a question about getting up-to-date residuals of `StressDivergenceTensor` Kernel using explicit solver e.g. `CentralDifference`.

I obtain the residuals through `save_in` option in `StressDivergenceTensor`:

```
[Modules]
    [./TensorMechanics]
      [./Master]
        [./all]
          strain = SMALL
          add_variables = true
          planar_formulation = PLANE_STRAIN
          generate_output = 'stress_xx stress_yy stress_xy'
          save_in = 'resid_x resid_y'
        [../]
      [../]
    [../]
  []
```

The residuals `resid_x resid_y` are then used as input parameters  in the custom cohesive zone `material kernel` along with `_interface_displacement_jump[_qp]`  to update the local `traction` vector.

The kernel gives acceleration is as follows:

```
[Kernels]
    [./inertia_x]
        type = InertialForce
        use_displaced_mesh = false
        variable = disp_x
    []
    [./inertia_y]
        type = InertialForce
        use_displaced_mesh = false
        variable = disp_y
    []
[]
```

In our current **traction-separation law**, shear stress `_stress[_qp](1)` combined with `_interface_displacement_jump[_qp]` gives **correct answer**. However, this is an approximation to the original theory and this "shear stress term" is needed be replaced by difference of reaction at upper/lower surface divided by area `(_resid_x[_qp] - _resid_x_neighbor[_qp])/(2*area)`, and that is where the problem comes in.

To better illustrate what is happening, here I showed the output (variables printed inside the `material kernel`) for the first several steps: 

```
Time Step 0, time = 0

Time Step 1, time = 0.0025, dt = 0.0025
displacement jump: 0
reaction: 0
shear stress: 0
 **Solve Converged!**
displacement jump: -3.37079e-05
reaction: 0
shear stress: -5400

Time Step 2, time = 0.005, dt = 0.0025
displacement jump: -3.37079e-05
reaction: 0
shear stress: -5400
 **Solve Converged!**
displacement jump: -0.000100711
reaction: -5400
shear stress: -16089.4

Time Step 3, time = 0.0075, dt = 0.0025
displacement jump: -0.000100711
reaction: 0
shear stress: -16089.4
 **Solve Converged!**
displacement jump: -0.000200246
reaction: -16089.4
shear stress: -31862
```

To be precise, `shear stress: ` represents shear stress term (which gives correct results), `reaction: ` represents the reaction difference term `(_resid_x[_qp] - _resid_x_neighbor[_qp])/(2*area)` (which uses the variable I extracted from `StressDivergenceTerm`. 

From the output we can see 

1. `Material kernel` is evaluated before/after solve.

2. `_interface_displacement_jump[_qp]` and `_stress[_qp](1)` are closely updated once the new displacements are available, which is the default feature of `material kernel`.

3. `reaction` only updates after solve (the `reaction: 0` can be solved by introducing `auxkernel` and save old values) but retrieve the residuals before the new displacements are calculated, in other words, `reaction` has time step offsets with `shear stress`.

I have searched several options, but currently none of them is working:

1.  Define `reaction` as `material property` same as `stress`. (But  `_grad_test[_i][_qp]` `computeQpResidual()` is not available in `material kernel`)

2.  Evaluate it again right after the new displacements are available. (Probably need a separate `kernel object` to achieve it or do multi-app, which is not trivial)

3.  Use `Tagging Interface` or `Control System` as suggested by similar posts (But I can't find a solution from them, because this seems to be issue of `save_in` which is not re-evaluating residuals after the solve )

I'm struggling to find a workaround of it. Any suggestion or comment is welcome. Thanks a lot!

**Comment by GiudGiud** [comment, marked as answer]
Hello

so you need the output from one kernel (the residual) to compute the residual from another kernel?

Guillaume

**Reply by chunhuizhao478** [reply]
Hi, basically I need the residual from `StressDivergenceKernel` as a variable to compute `traction` (not the `residual`) in `cohesive zone model material kernel`, but it seems like it is not up-to-date as I showed above.

**Reply by GiudGiud** [reply]
This is tough since because all the kernels are executed together.
Can you duplicate the calculation in StressDivergenceKernel and place it in a material property instead? Material properties are calculated before kernels

**Reply by chunhuizhao478** [reply]
I'm not sure whether I understand this correctly, is it fine to inherit all the necessary header files of `StressDivergenceKernel` and duplicate all the residual calculations but pass as material property by overriding `computeQpProperties()` in `material kernel`?

**Reply by GiudGiud** [reply]
No this wont work. An object may not be both a kernel and a material property.

You would have to create a material property and duplicate the code that computes the residual in the Kernel. 

Maybe there are other solutions, through the tagging system.
You would have to make sure one kernel gets executed before the other one though

**Reply by chunhuizhao478** [reply]
Thanks for your reply. I would like to try to define a material property first. Maybe we are referring to the same method (not trying to define kernel and material property at the same time). Let me try to explain my thought here: After looking through the code in `StressDivergenceKernel.C`, the code snippets for calculating residuals are the following (I would need to include the necessary header files though) :

```
void
StressDivergenceTensors::computeResidual()
{
  prepareVectorTag(_assembly, _var.number());

  if (_volumetric_locking_correction)
    computeAverageGradientTest();

  precalculateResidual();
  for (_i = 0; _i < _test.size(); ++_i)
    for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
      _local_re(_i) += _JxW[_qp] * _coord[_qp] * computeQpResidual();

  accumulateTaggedLocalResidual();

  if (_has_save_in)
  {
    Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
    for (const auto & var : _save_in)
      var->sys().solution().add_vector(_local_re, var->dofIndices());
  }
}

Real
StressDivergenceTensors::computeQpResidual()
{
  Real residual = _stress[_qp].row(_component) * _grad_test[_i][_qp];
  // volumetric locking correction
  if (_volumetric_locking_correction)
    residual += _stress[_qp].trace() / 3.0 *
                (_avg_grad_test[_i][_component] - _grad_test[_i][_qp](_component));

  if (_ndisp != 3 && _out_of_plane_strain_coupled && _use_displaced_mesh)
  {
    const Real out_of_plane_thickness = std::exp((*_out_of_plane_strain)[_qp]);
    residual *= out_of_plane_thickness;
  }

  return residual;
}

```
The quantity I would like to extract is `_local_re` after `accumulateTaggedLocalResidual()` and evaluate at each `_qp` point of interest, without considering the volumetric locking, I should write something like:
```
void
CompResidualCustom::computeQpProperties()
{
   computeResidualCustom();
    _local_re_out[_qp] = _local_re[_qp];
}

void
CompResidualCustom::computeResidualCustom()
{
  prepareVectorTag(_assembly, _var.number());

  precalculateResidual();
  for (_i = 0; _i < _test.size(); ++_i)
    for (_qp = 0; _qp < _qrule->n_points(); ++_qp)
      _local_re(_i) += _JxW[_qp] * _coord[_qp] * computeQpResidualCustom();

  accumulateTaggedLocalResidual();
}

Real
CompResidualCustom::computeQpResidualCustom()
{
  Real residual = _stress[_qp].row(_component) * _grad_test[_i][_qp];
  return residual;
}
```
not sure if it is doable to pass `_local_re[_qp]` inside  `computeQpProperties()` in `material kernel` or there is something else I need to be careful when trying to duplicate the calculation. Thanks!

**Reply by GiudGiud** [reply]
Yeah this seems difficult. The local weights are not available in a material property I think.

**Reply by chunhuizhao478** [reply]
Hi Guillaume, thanks for your reply, Just want to let you know, I manage to solve the problem using `multiapp`, I will add some comments about it later.

**Comment by hugary1995** [comment]
This is a real issue, and I agree with @GiudGiud this is very difficult. However, I don't think this has any serious practical consequences with explicit time integration, i.e., the response of the system (for instance the force displacement curve) won't be sensitive to the lagged residual.
